doc/rtd/reference/datasources/nocloud.rst

```diff
@@ -4,97 +4,237 @@ NoCloud
+We will describe the possible values in a line configuration in the following
+sections.
 
 .. note::
+
    If using kernel command line arguments with GRUB, note that an
    unescaped semicolon is intepreted as the end of a statement.
-   Consider using single-quotes to avoid this pitfall. See: `GRUB quoting`_
-   ds=nocloud;s=http://10.42.42.42/cloud-init/configs/
+   See: `GRUB quoting`_
+
+Method 2: System configuration
+------------------------------
+
+System configurations are YAML-formatted files and have names that end in
+``.cfg``. These are located under :file:`/etc/cloud/cloud.cfg.d/`.
+
+Example:
+
+.. code-block:: yaml
+
+   datasource:
+     NoCloud:
+       seedfrom: https://10.42.42.42/configs/
+
+The above system configuration tells cloud-init that it is using NoCloud and
+that it can find configurations at ``https://10.42.42.42/configs/``.
+
+The scope of this section is limited to its use for selecting the source of
+its configuration, however it is worth mentioning that the system configuration
+provides more than just the meta-configuration.
+
+In addition to defining where cloud-init can find runtime configurations, the
+system configuration also controls many of cloud-init's default behaviors.
+Most users shouldn't need to modify these defaults, however it is worth noting
+that downstream distributions often use them to set reasonable default
+behaviors for cloud-init. This includes things such as which distro to behave
+as and which networking backend to use.
+
+The default values in :file:`/etc/cloud/cloud.cfg` may be overriden by drop-in
+files which are stored in :file:`/etc/cloud/cloud.cfg.d`.
+
+Configuration sources
+=====================
+
+User-data, meta data, network config (optional), and vendor data
+(optional) may be sourced from one of several possible locations, either
+locally or remotely.
+
+Source 1: Local filesystem
+--------------------------
+
+System configuration may provide cloud-init runtime configuration directly
+
+.. code-block:: yaml
+
+   datasource:
+     NoCloud:
+       meta-data: |
+         instance-id: l-eadfbe
+       user-data: |
+         #cloud-config
+         runcmd: [ echo "it worked!" > /tmp/example.txt ]
 
-Alternatively, this URI may be defined in a configuration in a file
-:file:`/etc/cloud/cloud.cfg.d/*.cfg` like this: ::
+Local filesystem: custom location
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+Cloud-init makes it possible to find system configuration in a custom
+filesystem path for those that require more flexibility. This may be
+done with a line configuration: ::
+
+
+  ds=nocloud;s=file://path/to/directory/
+
+Or a system configuration:
+
+.. code-block:: yaml
+
+   datasource:
+     NoCloud:
+       seedfrom: file://path/to/directory
+
+Source 2: Drive with labeled filesystem
+---------------------------------------
+
+A labeled `vfat`_ or `iso9660` filesystem may be used. The filesystem volume
+must be labelled ``CIDATA``. The configuration files must be in the root
```

Comment:
What are the expected / allowed names of those files? Are they the same as in `reference/datasources/nocloud.html#source-files`?